### PR TITLE
Fix casting issue when using var with a table returned from a query

### LINKED ISSUE #28387: Bad type on operand stack error when using var with a table returned from a query
**Description:**
$title.

**Steps to reproduce:**
```ballerina
type Employee record {
    readonly string name;
    int id;
};

public function main() {
    table<Employee> key(name) t = table [
        {name: "Amy", id: 1234},
        {name: "John", id: 4567}
    ];

    var ids =
        from var { id } in t
        select {id}; // or `select id`
}
```

```log
[2021-02-02 17:20:23,356] SEVERE {b7a.log.crash} - Bad type on operand stack
Exception Details:
  Location:
    xyz.main(Lio/ballerina/runtime/internal/scheduling/Strand;)Ljava/lang/Object; @4103: putfield
  Reason:
    Type 'java/lang/Object' (current frame, stack[2]) is not assignable to 'io/ballerina/runtime/internal/values/TableValueImpl'
  Current Frame:
    bci: @4103
    flags: { }
    locals: { 'io/ballerina/runtime/internal/scheduling/Strand', 'io/ballerina/runtime/internal/values/TableValueImpl', 'io/ballerina/runtime/internal/values/TableValueImpl', 'io/ballerina/runtime/internal/values/ArrayValue', long, long_2nd, 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/internal/values/ArrayValue', 'io/ballerina/runtime/internal/values/TypedescValue', 'java/lang/Object', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/api/values/BString', long, long_2nd, 'java/lang/Object', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/api/values/BString', 'io/ballerina/runtime/api/values/BString', long, long_2nd, 'java/lang/Object', 'io/ballerina/runtime/internal/values/TableValueImpl', 'io/ballerina/runtime/api/values/BObject', 'io/ballerina/runtime/internal/values/TypedescValue', 'io/ballerina/runtime/api/values/BObject', 'io/ballerina/runtime/api/values/BObject', 'io/ballerina/runtime/internal/values/FPValue', 'java/lang/Object', 'io/ballerina/runtime/api/values/BObject', 'io/ballerina/runtime/internal/values/FPValue', 'java/lang/Object', 'io/ballerina/runtime/internal/values/StreamValue', 'io/ballerina/runtime/internal/values/StreamValue', 'io/ballerina/runtime/internal/values/TableValueImpl', 'io/ballerina/runtime/internal/values/ArrayValue', 'java/lang/Object', 'java/lang/Object', 'java/lang/Object', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'io/ballerina/runtime/internal/values/ErrorValue', integer, 'io/ballerina/runtime/internal/values/ErrorValue', 'java/lang/Object', integer }
    stack: { 'mainFrame', 'mainFrame', 'java/lang/Object' }
```

**Affected Versions:**
slalpha1

### LINKED ISSUE #28397: VSCode doesn't provide suggestions after some time
Related to #28396

Issue 1: I am getting completely unrelated diagnostics now 
Screen Shot 2021-02-02 at 1 51 10 PM

Screen Shot 2021-02-02 at 1 49 38 PM

Issue 2: Now I am not getting any suggestions
Screen Shot 2021-02-02 at 1 51 43 PM

No logs in the output tab

### LINKED ISSUE #30100: Semantic Analyzer fails with CCE
**Description:**
Consider the following source

```
type TestT int & string;

type TypeName int;
```

This is failing with the following stack-trace.

```
'class org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.symbols.BTypeSymbol (org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol and org.wso2.ballerinalang.compiler.semantics.model.symbols.BTypeSymbol are in unnamed module of loader 'app')'}
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.symbols.BTypeSymbol (org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol and org.wso2.ballerinalang.compiler.semantics.model.symbols.BTypeSymbol are in unnamed module of loader 'app')
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.invalidateAlreadyDefinedErrorType(SymbolEnter.java:1499)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:1372)
	at org.wso2.ballerinalang.compiler.tree.BLangTypeDefinition.accept(BLangTypeDefinition.java:111)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:270)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineTypeNodes(SymbolEnter.java:1105)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineConstructs(SymbolEnter.java:379)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:313)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.defineNode(SymbolEnter.java:270)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.definePackage(SymbolEnter.java:263)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:377)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
	at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:176)
	at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:168)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:97)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
	at io.ballerina.projects.Package.getCompilation(Package.java:131)
	at org.ballerinalang.langserver.workspace.BallerinaWorkspaceManager.waitAndGetPackageCompilation(BallerinaWorkspaceManager.java:218)
	at org.ballerinalang.langserver.diagnostic.DiagnosticsHelper.getLatestDiagnostics(DiagnosticsHelper.java:115)
	at org.ballerinalang.langserver.diagnostic.DiagnosticsHelper.compileAndSendDiagnostics(DiagnosticsHelper.java:80)
	at org.ballerinalang.langserver.BallerinaTextDocumentService.didChange(BallerinaTextDocumentService.java:538)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Affected Versions:**
sl-alpha4 at least

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fix following issues
- Fix casting issue when using var with a table returned from a query
- Fix CCE issue with type definitions

Fixes #28387 
Fixes #30100 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal [hunk 1]
```diff
@@ -354,3 +354,24 @@ function testVariableShadowingWithQueryExpressions2() returns boolean {
...
+    int id;
+};
+
+function testSimpleSelectQueryWithTable() {
+    table<Student> key(name) t = table [
+        {name: "Amy", id: 1234},
+        {name: "John", id: 4567}
+    ];
+    var ids = from var { id } in t select {id};
+    assertEquality(2, ids.length());
```
Comment (grainier) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal:369]: Shall we assert array content as well?

**Review by grainier — APPROVED** [review]
LGTM

File: tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal [hunk 2]
```diff
@@ -354,3 +354,24 @@ function testVariableShadowingWithQueryExpressions2() returns boolean {
...
+    readonly string name;
+    int id;
+};
+
+function testSimpleSelectQueryWithTable() {
+    table<Student> key(name) t = table [
+        {name: "Amy", id: 1234},
+        {name: "John", id: 4567}
+    ];
+    var ids = from var { id } in t select {id};
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/simple-query-with-var-type.bal:368]:
```suggestion
    var ids = from var {id} in t select {id};
```

File: tests/jballerina-unit-test/src/test/resources/test-src/typedefs/type-definitions-semantics-negative.bal [hunk 3]
```diff
@@ -16,3 +16,5 @@ type T8 [int[], A[], [B, C], map<string>, map<D>, E, int, record { F f; }, objec
 
 type PersonOrInt int|Person;
 type PersonOrNil Person?;
+
+type TestT int & string;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/typedefs/type-definitions-semantics-negative.bal:20]: Shall we use a name like `InvalidIntersectionType`?

**Review by mohanvive — APPROVED** [review]

**Pull request merged**